src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,219 @@
+
+namespace Microsoft.CodeAnalysis.BraceCompletion
+{
+    internal abstract class AbstractBraceCompletionService : IBraceCompletionService
+    {
+        protected abstract char OpeningBrace { get; }
+
+        protected abstract char ClosingBrace { get; }
+
+        /// <summary>
+        /// Returns if the token is a valid opening token kind for this brace completion service.
+        /// </summary>
+        protected abstract bool IsValidOpeningBraceToken(SyntaxToken token);
+
+        /// <summary>
+        /// Returns if the token is a valid closing token kind for this brace completion service.
+        /// </summary>
+        protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
+
+        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
+
+        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            var closingPoint = braceCompletionContext.ClosingPoint;
+            if (closingPoint < 1)
+            {
+                return null;
+            }
+
+            var openingPoint = braceCompletionContext.OpeningPoint;
+            var document = braceCompletionContext.Document;
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (sourceText[openingPoint] != OpeningBrace)
+            {
+                return null;
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(openingPoint, findInsideTrivia: true);
+            var validOpeningPoint = await IsValidOpenBraceTokenAtPositionAsync(token, openingPoint, document, cancellationToken).ConfigureAwait(false);
+            if (!validOpeningPoint)
+            {
+                return null;
+            }
+
+            var braceTextEdit = new TextChange(TextSpan.FromBounds(closingPoint, closingPoint), ClosingBrace.ToString());
+            // The caret location should be in between the braces.
+            var originalOpeningLinePosition = sourceText.Lines.GetLinePosition(openingPoint);
+            var caretLocation = new LinePosition(originalOpeningLinePosition.Line, originalOpeningLinePosition.Character + 1);
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation);
+        }
+
+        public virtual Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
```

Comment:
hrmm... why are thse both virtual?  woudl it be better to make each subclass have to think about this?  or is the general idea that each subclass implements at least one of these?

Reply:
In almost all cases the sub classes don't override this - they don't do anything special.  The only one that does is curly brace completion